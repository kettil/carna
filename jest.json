{
  "globals": {
    "ts-jest": {
      "tsConfig": "tsconfig.json",
      "diagnostics": {
        "warnOnly": true,
        "importHelpers": true
      }
    }
  },

  "preset": "ts-jest",
  "testMatch": ["**/*.test.{js,jsx,ts,tsx}"],
  "roots": ["<rootDir>/src/"],

  "clearMocks": true,
  "restoreMocks": true,

  "collectCoverageFrom": ["src/**/*.{ts,tsx,js,jsx}"],
  "coveragePathIgnorePatterns": ["/__jest__/", "types.ts"],
  "coverageReporters": ["text-summary", "html"],
  "coverageDirectory": "coverage",
  "coverageThreshold": {
    "global": {
      "statements": 100,
      "branches": 100,
      "functions": 100,
      "lines": 100
    }
  }
}
