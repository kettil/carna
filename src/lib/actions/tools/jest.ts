import { join } from 'path';
import { jestCommand, jestConfigFiles } from '../../../configs/actionConfigs';
import { ExecOptions } from '../../cmd/exec';
import { execLog } from '../../cmd/execLog';
import { execStdio } from '../../cmd/execStdio';
import { Action } from '../../types';
import { getCoverageFolder } from '../../utils/getCoverageFolder';
import { getFirstExistingFile } from '../../utils/getFirstExistingFile';
import { JestActionProps } from '../types';

const options: Array<keyof JestActionProps> = ['updateSnapshot', 'runInBand', 'watch', 'verbose'];

const jestAction: Action<JestActionProps> = async ({ root, ci, log }, props) => {
  const coverageFolder = getCoverageFolder(root, props.projects);
  const configFile = await getFirstExistingFile({ cwd: root, files: jestConfigFiles });

  const cmd = join(root, jestCommand);
  const args: string[] = ['--config', configFile, '--colors'];

  if (props.projects.length === 1) {
    args.push('--selectProjects', props.projects[0]);
  } else if (props.projects.length > 1) {
    args.push('--selectProjects', ...props.projects);
  }

  options.forEach((option) => {
    if (props[option] === true) {
      args.push(`--${option}`);
    }
  });

  if (!props.runInBand) {
    // see https://ivantanev.com/make-jest-faster
    if (props.watch) {
      args.push('--maxWorkers', '25%');
    } else {
      args.push('--maxWorkers', '50%');
    }
  }

  if (ci) {
    args.push('--ci');
    args.push('--silent');
    args.push('--json');
    args.push('--outputFile', join(coverageFolder, 'jest-final.json'));
  }

  args.push('--coverage');
  args.push('--coverageDirectory', coverageFolder);
  args.push('--coverageReporters', 'json');
  args.push('--coverageThreshold', '\'{"global":{"statements":0,"branches":0,"functions":0,"lines":0}}\'');

  if (props.watch) {
    args.push('--coverageReporters', 'json-summary');
    args.push('--coverageReporters', 'lcovonly');
    args.push('--coverageReporters', 'html');
  }

  log.info('Run test');

  const execOptions: ExecOptions = { cmd, args, cwd: root, log };

  if (props.watch) {
    await execStdio(execOptions, { registerStdin: true });
  } else {
    await execLog(execOptions);
  }
};

export { jestAction };
