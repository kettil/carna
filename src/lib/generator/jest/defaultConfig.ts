const defaultConfig = {
  bail: 10,

  testEnvironment: 'node',

  clearMocks: true,
  restoreMocks: true,
} as const;

export { defaultConfig };
