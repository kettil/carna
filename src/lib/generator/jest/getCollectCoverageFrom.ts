import { isArray } from '@kettil/tool-lib';

const getCollectCoverageFrom = (workspaces: string[], customeCollectCoverageFrom?: unknown): string[] => {
  const workspacePaths = workspaces.flatMap((workspace) => [
    `${workspace}/src/**/*.{ts,tsx,js}`,
    `!${workspace}/src/**/types.ts`,
    `!${workspace}/src/index.ts`,
  ]);

  return [
    '<rootDir>/src/**/*.{ts,tsx,js}',
    '!<rootDir>/src/**/types.ts',
    '!<rootDir>/src/index.ts',

    ...workspacePaths,
    ...(isArray(customeCollectCoverageFrom)
      ? customeCollectCoverageFrom.filter((value): value is string => typeof value === 'string')
      : []),
  ];
};

export { getCollectCoverageFrom };
