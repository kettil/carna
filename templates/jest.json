{
  "globals": {
    "ts-jest": {
      "tsConfig": "tsconfig.json",
      "diagnostics": {
        "warnOnly": true,
        "importHelpers": true
      }
    }
  },

  "preset": "ts-jest",
  "testMatch": ["**/*.test.{js,jsx,ts,tsx}"],
  "roots": ["<rootDir>/src/"],

  "clearMocks": true,
  "restoreMocks": true,

  "collectCoverageFrom": ["src/**/*.{ts,tsx,js,jsx}"],
  "coveragePathIgnorePatterns": ["/__jest__/", "types.ts"],
  "coverageReporters": ["text-summary", "html"],
  "coverageDirectory": "coverage",
  "coverageThreshold": {
    "global": {
      "statements": 90,
      "branches": 90,
      "functions": 90,
      "lines": 90
    }
  }
}
