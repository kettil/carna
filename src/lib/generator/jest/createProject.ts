import { join } from 'path';
import { isObject, isString } from '@kettil/tool-lib';
import { defaultConfig } from './defaultConfig';
import { getSetupFilesAfterEnvironment } from './getSetupFilesAfterEnvironment';

const createProject =
  (settings: unknown) =>
    (root: string, workspaces: Record<string, string>): Record<string, unknown> => {
      if (!isObject(settings)) {
        throw new TypeError('createProject requires an object as the first argument');
      }

      const { testFolder, workspaceFolder, color, customeConfig = {} } = settings;
      const hasWorkspace = isString(workspaceFolder);

      if (hasWorkspace && !Object.keys(workspaces).includes(workspaceFolder)) {
        throw new TypeError(`Workspace "${workspaceFolder}" is unknown - createProject({ workspaceFolder: "<name>" })`);
      }

      if (typeof testFolder !== 'string') {
        throw new TypeError('"testFolder" is not a string - createProject({ testFolder: "<name>" })');
      }

      if (typeof color !== 'string') {
        throw new TypeError('"color" is not a color string - createProject({ color: "<color>" })');
      }

      if (!isObject(customeConfig)) {
        throw new TypeError('"customeConfig" is not an object - createProject({ customeConfig: { ... } })');
      }

      const pathWorkspace = hasWorkspace ? workspaces[workspaceFolder] : '';

      const configPathSource = hasWorkspace ? `${pathWorkspace}/src` : '<rootDir>/src';
      const configPathTests = hasWorkspace ? `${pathWorkspace}/tests/${testFolder}` : `<rootDir>/tests/${testFolder}`;

      const name = (hasWorkspace ? `${workspaceFolder}:` : '') + testFolder;

      const pathTest = join(root, configPathTests.replace('<rootDir>', '.'));

      return {
        ...defaultConfig,

        displayName: { name, color },
        roots: [configPathSource, configPathTests],
        testMatch: ['**/*.test.{js,ts,tsx}'],
        testPathIgnorePatterns: ['/node_modules/', '/src/'],
        setupFilesAfterEnv: getSetupFilesAfterEnvironment(pathTest, ['setupTests']),

        ...customeConfig,
      };
    };

export { createProject };
