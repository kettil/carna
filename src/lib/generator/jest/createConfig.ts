import { isArray, isObject } from '@kettil/tool-lib';
import { defaultConfig } from './defaultConfig';
import { getCollectCoverageFrom } from './getCollectCoverageFrom';
import { getWorkspaces } from './getWorkspaces';

const createConfig = (root: string, config: unknown): Record<string, unknown> => {
  if (!isObject(config)) {
    throw new TypeError('"config" is not an object');
  }

  if (!isArray(config.projects)) {
    throw new TypeError('"projects" is not an array in the "config" object');
  }

  const workspaces = getWorkspaces(root);
  const collectCoverageFrom = getCollectCoverageFrom(Object.values(workspaces), config.collectCoverageFrom);
  const projects = config.projects
    .filter(
      (project): project is Record<string, unknown> | ((...args: unknown[]) => unknown) =>
        isObject(project) || typeof project === 'function',
    )
    .map((project) => (typeof project === 'function' ? project(root, workspaces) : { ...defaultConfig, ...project }));

  return {
    ...defaultConfig,
    ...config,

    projects,
    collectCoverageFrom,
  };
};

export { createConfig };
