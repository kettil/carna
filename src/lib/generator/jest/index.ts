import { createConfig } from './createConfig';
import { createProject } from './createProject';

export { createConfig as createJestConfig, createProject as createJestProject };
