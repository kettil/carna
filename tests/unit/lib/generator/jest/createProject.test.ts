import { createProject } from '../../../../../src/lib/generator/jest/createProject';

describe('createProject()', () => {
  test('it should work without workspaces', () => {
    const settings = {
      testFolder: 'unit',
      color: 'green',
    };
    const workspaces = {};

    const result = createProject(settings)('/path/to/project', workspaces);

    expect(result).toMatchSnapshot();
  });

  test('it should work with workspaces', () => {
    const settings = {
      testFolder: 'unit',
      color: 'green',
      workspaceFolder: 'foo',
    };
    const workspaces = {
      foo: '/path/to/project/packages/foo',
      bar: '/path/to/project/packages/bar',
    };

    const result = createProject(settings)('/path/to/project', workspaces);

    expect(result).toMatchSnapshot();
  });

  test('it should thrown an error', () => {
    const settings = undefined;

    const callbackWithException = () => createProject(settings)('/path/to/project', {});

    expect(callbackWithException).toThrow('createProject requires an object as the first argument');
  });
});
