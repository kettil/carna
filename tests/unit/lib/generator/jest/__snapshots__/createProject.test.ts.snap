// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createProject() it should work with workspaces 1`] = `
Object {
  "bail": 10,
  "clearMocks": true,
  "displayName": Object {
    "color": "green",
    "name": "foo:unit",
  },
  "restoreMocks": true,
  "roots": Array [
    "/path/to/project/packages/foo/src",
    "/path/to/project/packages/foo/tests/unit",
  ],
  "setupFilesAfterEnv": Array [],
  "testEnvironment": "node",
  "testMatch": Array [
    "**/*.test.{js,ts,tsx}",
  ],
  "testPathIgnorePatterns": Array [
    "/node_modules/",
    "/src/",
  ],
}
`;

exports[`createProject() it should work without workspaces 1`] = `
Object {
  "bail": 10,
  "clearMocks": true,
  "displayName": Object {
    "color": "green",
    "name": "unit",
  },
  "restoreMocks": true,
  "roots": Array [
    "<rootDir>/src",
    "<rootDir>/tests/unit",
  ],
  "setupFilesAfterEnv": Array [],
  "testEnvironment": "node",
  "testMatch": Array [
    "**/*.test.{js,ts,tsx}",
  ],
  "testPathIgnorePatterns": Array [
    "/node_modules/",
    "/src/",
  ],
}
`;
